gold/Gold_core.hpp

```diff
@@ -0,0 +1,156 @@
+
+#pragma once
+
+#include <deque>
+
+#include "explicit_type.hpp"
+#include "Gold_data.hpp"
+#include "Gold_mem.hpp"
+
+using Inst_id = Explicit_type<int, struct Inst_id_struct, 0>;
+
+
+enum class Mem_op {
+  Invalid,
+  Load,
+  Store,
+  Ack,
+  Rel,
+  AckRel
+};
+
+class Gold_core {
+public:
+  /** inorder
+   * Gets the in-order ID for instructions through the pipeline
+   */
+  Inst_id inorder();
+
+  /** set_type
+   * Sets the type of operation. Must be called before set_safe, some checks
+   * are dependent on the memory oeration
+   */
+  void set_type(Inst_id iid, Mem_op op);
+
+  /** set_safe
+   * When the memory operation can not be undone (no possible to nuke afterwards)
```

Comment:
You mean the all operations prior to and including the one designated by iid cannot be undo?